Source/Model/Message/ZMClientMessage.swift

```diff
@@ -0,0 +1,173 @@
+            // This way if we get a delete from a different device while we are waiting for the response it will delete this message
+            let originalID = self.genericMessage.flatMap { UUID(uuidString: $0.edited.replacingMessageId) }
+            self.nonce = originalID
+        }
+        super.expire()
+    }
+
+    public override func resend() {
+        if let genericMessage = self.genericMessage, genericMessage.hasEdited() {
+            // Re-apply the edit since we've restored the orignal nonce when the message expired
+            editText(self.textMessageData?.messageText ?? "", mentions: self.textMessageData?.mentions ?? [], fetchLinkPreview: true)
+        }
+        super.resend()
+    }
+    
+    public override func update(withPostPayload payload: [AnyHashable : Any], updatedKeys: Set<AnyHashable>?) {
+        // we don't want to update the conversation if the message is a confirmation message
+        guard let genericMessage = self.genericMessage else { return }
+        if genericMessage.hasConfirmation() || genericMessage.hasReaction() {
+            return
+        }
+
+        if genericMessage.hasDeleted() {
+            let originalID = UUID(uuidString: genericMessage.deleted.messageId)
+            guard let managedObjectContext = managedObjectContext,
+                let conversation = conversation else { return }
+
+            let original = ZMMessage.fetch(withNonce: originalID, for: conversation, in: managedObjectContext)
+            original?.sender = nil
+            original?.senderClientID = nil
+        } else if genericMessage.hasEdited() {
+            if let nonce = self.nonce(fromPostPayload: payload),
+                self.nonce != nonce {
+                ZMSLog(tag: "send message response nonce does not match")
+                return
+            }
+
+            if let serverTimestamp = (payload as NSDictionary).optionalDate(forKey: "time") {
+                self.updatedTimestamp = serverTimestamp
+            }
+        } else {
+            super.update(withPostPayload: payload, updatedKeys: nil)
+        }
+    }
+
+    override static public func predicateForObjectsThatNeedToBeInsertedUpstream() -> NSPredicate? {
+        let encryptedNotSynced = NSPredicate(format: "%K == FALSE", DeliveredKey)
+        let notExpired = NSPredicate(format: "%K == 0", ZMMessageIsExpiredKey)
+        return NSCompoundPredicate(andPredicateWithSubpredicates: [encryptedNotSynced, notExpired])
+    }
+    
+    public override func markAsSent() {
+        super.markAsSent()
+        if linkPreviewState == ZMLinkPreviewState.uploaded {
+            linkPreviewState = ZMLinkPreviewState.done
+        }
+        setObfuscationTimerIfNeeded()
+    }
+
+    private func setObfuscationTimerIfNeeded() {
+        guard self.isEphemeral else {
+            return
+        }
+        if let genericMessage = self.genericMessage,
+            let _ = genericMessage.textData,
```

Comment:
```suggestion
            genericMessage.textData != nil,
```

In my opinion this way it's more clear, that we are checking that the property is not nil.